codemodder/codemods/transformations/clean_imports.py

```diff
@@ -57,148 +57,272 @@ def __init__(self, context: CodemodContext, src_path):
                 )
             )
+        return new_from_import_stmts
+
+    def _order_block(self, imports_block):
+        # triage from imports from regular imports
+        from_imports = []
+        regular_imports = []
+        for stmt in imports_block:
+            imp = stmt.body[0]
+            if matchers.matches(imp, matchers.ImportFrom()):
+                from_imports.append(stmt)
+            if matchers.matches(imp, matchers.Import()):
+                regular_imports.append(stmt)
+
+        # create the new statements to be inserted accoding to cases
+        regular_import_stmts = self._handle_regular_imports(regular_imports)
+        from_import_stmts = self._handle_from_imports(from_imports)
 
         # classify by sections and sort each section by module name
         imports_by_section = _triage_imports(
-            itertools.chain(module_imports_set, from_imports_set),
+            regular_import_stmts,
             self.src_path,
         )
         _sort_each_section(imports_by_section)
 
-        # wrap all the imports into statements, remove empty sections
-        all_import_statement = [
-            [cst.SimpleStatementLine(body=[imp]) for imp in section]
-            for section in imports_by_section
-            if section
-        ]
-
-        # add empty lines after each section
-        for section in all_import_statement[:-1]:
-            if section:
-                section.append(cst.EmptyLine())
-
-        # rewrite all the imports, tree may be empty
-        if result_tree.children:
-            all_import_statement.append(
-                [
-                    cst.EmptyLine(),
-                    result_tree.children[0].with_changes(leading_lines=[]),
-                ]
-            )
-            sentinel = cst.FlattenSentinel(
-                itertools.chain.from_iterable(all_import_statement)
-            )
-            result_tree = result_tree.deep_replace(result_tree.children[0], sentinel)
-            return result_tree
-        return result_tree.with_changes(
-            body=list(itertools.chain.from_iterable(all_import_statement))
+        from_imports_by_section = _triage_imports(
+            from_import_stmts,
+            self.src_path,
         )
+        _sort_each_section(from_imports_by_section)
 
+        # merge regular and from imports and filter empty sections
+        for i, section in enumerate(imports_by_section):
+            section.extend(from_imports_by_section[i])
 
-class RemoveNodes(CSTTransformer):
-    """
-    Remove given nodes from the tree.
-    """
+        imports_by_section = list(filter(lambda x: x, imports_by_section))
 
-    def __init__(self, nodes_to_remove: Set[cst.CSTNode]):
-        self.nodes_to_remove = nodes_to_remove
+        # add empty lines after each section
+        # for this, we add an empty line at the leading_lines of the first not at each section
+        for section in imports_by_section[1:]:
```

Comment:
The `[1:]` here is a bit magical and it's going to be hard to remember the rationale for this in the future.

Reply:
The comment could be a bit more clear, but essentially instead of adding an `EmptyLine` node right after the last node of each section, we're adding the node at the `leading_lines` at the first node of each section (thus the first one is skipped).

Reason being is that, in libcst, nodes with formatting information are usually owned by the node right after it. This way we avoid reporting possible changes when no visual changes happens.